prow/tide/codereview.go

```diff
@@ -161,3 +170,151 @@ func CodeReviewCommonFromPullRequest(pr *PullRequest) *CodeReviewCommon {
+	refsForJob(sp subpool, prs []CodeReviewCommon) prowapi.Refs
+	prMergeMethod(crc *CodeReviewCommon) (types.PullRequestMergeType, error)
+}
+
+// GitHubInteractor implements provider, used by tide Controller for
+// interacting directly with GitHub.
+//
+// Tide Controller should only use GitHubInteractor for communicating with GitHub.
+type GitHubInteractor struct {
+	cfg                config.Getter
+	ghc                githubClient
+	usesGitHubAppsAuth bool
+
+	*mergeChecker
+	logger *logrus.Entry
+}
+
+func (gi *GitHubInteractor) blockers() (blockers.Blockers, error) {
+	label := gi.cfg().Tide.BlockerLabel
+	if label == "" {
+		return blockers.Blockers{}, nil
+	}
+
+	gi.logger.WithField("blocker_label", label).Debug("Searching for blocker issues")
+	orgExcepts, repos := gi.cfg().Tide.Queries.OrgExceptionsAndRepos()
+	orgs := make([]string, 0, len(orgExcepts))
+	for org := range orgExcepts {
+		orgs = append(orgs, org)
+	}
+	orgRepoQuery := orgRepoQueryStrings(orgs, repos.UnsortedList(), orgExcepts)
+	return blockers.FindAll(gi.ghc, gi.logger, label, orgRepoQuery, gi.usesGitHubAppsAuth)
+}
+
+// Query gets all open PRs based on tide configuration.
+func (gi *GitHubInteractor) Query() (map[string]CodeReviewCommon, error) {
+	lock := sync.Mutex{}
+	wg := sync.WaitGroup{}
+	prs := make(map[string]CodeReviewCommon)
+	var errs []error
+	for i, query := range gi.cfg().Tide.Queries {
+
+		// Use org-sharded queries only when GitHub apps auth is in use
+		var queries map[string]string
+		if gi.usesGitHubAppsAuth {
+			queries = query.OrgQueries()
+		} else {
+			queries = map[string]string{"": query.Query()}
+		}
+
+		for org, q := range queries {
+			org, q, i := org, q, i
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				results, err := gi.search(gi.ghc.QueryWithGitHubAppsSupport, gi.logger, q, time.Time{}, time.Now(), org)
+
+				resultString := "success"
+				if err != nil {
+					resultString = "error"
+				}
+				tideMetrics.queryResults.WithLabelValues(strconv.Itoa(i), org, resultString).Inc()
+
+				lock.Lock()
+				defer lock.Unlock()
+				if err != nil && len(results) == 0 {
+					gi.logger.WithField("query", q).WithError(err).Warn("Failed to execute query.")
+					errs = append(errs, fmt.Errorf("query %d, err: %w", i, err))
+					return
+				}
+				if err != nil {
+					gi.logger.WithError(err).WithField("query", q).Warning("found partial results")
+				}
+
+				for _, pr := range results {
+					crc := CodeReviewCommonFromPullRequest(&pr)
+					prs[prKey(crc)] = *crc
+				}
+			}()
+		}
+	}
+	wg.Wait()
+
+	return prs, utilerrors.NewAggregate(errs)
+}
+
+func (gi *GitHubInteractor) GetRef(org, repo, ref string) (string, error) {
+	return gi.ghc.GetRef(org, repo, ref)
+}
+
+// contextRequired is for tide to work with Gerrit, it's meant to always return
+// false so that `optional: ` and `report: ` are respected.
+func (gi *GitHubInteractor) contextRequired(*config.JobBase) bool {
```

Comment:
Shouldn't this be checked through the `contextChecker`? And maybe leave out gerrit additions for now to keep this a refactor only?